vital/types/geo_offset.h

```diff
@@ -0,0 +1,210 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief This file contains the interface to a local geographic offset coordinate system.
+ */
+
+#ifndef KWIVER_VITAL_GEO_OFFSET_H_
+#define KWIVER_VITAL_GEO_OFFSET_H_
+
+#include <vital/types/point.h>
+#include <vital/types/geo_point.h>
+
+namespace kwiver {
+namespace vital {
+
+// ----------------------------------------------------------------------------
+/** Geo-coordinate.
+ *
+ * This class represents a geolocated cartesian coordinate system 
+ * centered at any origin specified by the application.
+ */
+class VITAL_EXPORT geo_offset : public point_3d
+{
+public:
+  typedef Eigen::Matrix< double, 3, 1 > vector_type;
+
+  geo_offset()
+  {
+    m_value = vector_type::Zero();
+  }
+  geo_offset(vector_type const& v) { m_value=v; }
+
+  virtual ~geo_offset() = default;
+
+  geo_point_sptr& origin();
+  const geo_point_cptr origin() const;
+
+  /**
+   * \brief Get the WGS84 coordinates of the cartesian coordinates.
+   *
+   * \returns The WGS84 lon/lat/alt of the cartesian coordinates.
+   */
+  vector_type get_lon_lat_alt() const;
+  /**
+   * \brief Set the cartesian coordinates based on 2 geo_points
+   */
+  void set_from_geo_points(const geo_point_sptr origin, const geo_point_sptr location);
+
+
+protected:
+
+  geo_point_sptr origin_;
+
+};
+
+// ----------------------------------------------------------------------------
+/** Local Cartesian Conversion Utility.
+ *
+ * Base on the NGA GeoTrans library
+ * https://earth-info.nga.mil/GandG/update/index.php?action=home
+ *
+ * This class is a cleaned up version of the LocalCartesian and Geocentric classes provided in GeoTrans
+ * This allows a user to define a local cartesian coordinate system with any origin (expressed in WGS84)
+ */
+class VITAL_EXPORT local_cartesian
```

Comment:
This seems very closely related to the local_geo_cs class that already exists.  I don't think we should duplicate that functionality under a different name.

Reply:
local_cartesian is just a class containing the conversion code between a cartesian x,y,z point and wgs84 lla, given ANY geo_point as the cartesian origin. Currently local_geo_cs uses UTM as its basis for the cartesian system, but that is problematic if scene's cross the utm zone you are in (among other things) proj does not support this conversion where the cartesian origin is user specified. (That I could find.) This math supports it and seems very useful in kwiver, I think local_geo_cs should hold an instance of local_cartesian rather than a geo_point and use it in manipulating camera positions (not sure local_geo_cs is the right name for a class with functions tailor to specific vital type.

Reply:
This class and `local_geo_cs` serve the same purpose.  It would be confusing to have both.  This class is doing the right thing in terms of a true local tangent plane, however I like the interface of `local_geo_cs` better.  I would like to see the `local_geo_cs` classed updated to use this computation under the hood instead of the UTM approximation.  `local_geo_cs` and `local_cartesian` should be one in the same.  I'm okay with changing the name (maybe `local_cartesian_cs`?), but I don't understand what that last comment about specific vital types means.  If you mean the camera updating functions, yes those should be moved out of this class.

This class should still use a `geo_point` to specify the origin.  Or at set and access the origin as a `geo_point`.

Reply:
note local_cartesian does hold a geo_point member as the origin. 

Seems that it would be good to have 2 classes one that only uses Eigen types and one that takes in and operates on vital types. But if we want to have a big class with lots of methods of mixed types (some vector3 based methods, some camera based methods) that's fine too. 

I am fine with how ever you want to merge or refactor these two classes. But should probably be done on another PR?

Reply:
I'm not sure what you mean.  Most vital types are Eigen types or contain Eigen types.  I think it is a very bad idea to have two classes that appear to have the same function but are implemented in different ways and operate on different subsets of kwiver types.  In my opinion, the member functions of this class should operate only on fundamental types like vector and matrix.  Transformation of higher level types like camera and tracks should be external to the class.

My concern is that it we introduce this class now with the plan to refactor later, we will never get around to that refactoring and we will have code divergence and confusion about which class to use.  I'd rather not see the introduction of a "duplicate" class in this branch.  That said, I don't want to hold up the other work in this branch.  You could remove this class from this branch (use `local_geo_cs` for now if needed) and then start a new branch to upgrade/rename `local_geo_cs` using this code.

Reply:
I guess I think about the local_cartesian class as a utility much like the methods in the rotation.cxx 
And as a supplement to the coordinate transformations provided by the geo_point.location conversion function (so a compliment to proj)

local_geo_gs only has 2 methods on it that specifically take in a simple_camera_perspective and a metadata object, and then operates on the camera location and rotation matrix.
It uses the rotation methods defined in rotation.cxx to do the rotation calculations
And it uses the geo_point methods to do the translations methods

Currently, it uses geo_point.location to get the camera cartesian location using a UTM based position to offset

I thought instead, the local_geo_cs class would not hold a geo_point, but a local_cartesian object, and you would use the local_cartesian class to get the cartesian position of the camera based on the md_sensor_location

basically just changing these lines 
https://github.com/Kitware/kwiver/blob/dev/object_track_additions/vital/types/local_geo_cs.cxx#L144
https://github.com/Kitware/kwiver/blob/dev/object_track_additions/vital/types/local_geo_cs.cxx#L178
https://github.com/Kitware/kwiver/blob/dev/object_track_additions/vital/types/local_geo_cs.cxx#L314

local_cartesian does not do rotations nor understand what a camera or metadata is.
Its just translation math that's not provided in proj

If it makes sense to make the local_cartesian to hold the 'convert' methods and to add methods to take in classes like camera and metadata (landmarks and features too?) and have it update rotations as well, i am ok with that, it just seems to be putting a lot of functionality in one class

Sorry for being thick headed about this

Reply:
I'm sorry, I'm asking for too much on this branch.  `local_geo_cs` is really the class that need to be refactored and ultimately removed.  It was hastily ported from TeleSculptor but is really not a good fit as a vital type.  This class is much closer to what I originally intended `local_geo_cs` to be.  I think if we can do a few things to make this "compatible" with `local_geo_cs` then we can worry about converting users of `local_geo_cs` to use `local_cartesian` on another branch.  I'll put aside my worries that we will forget to do the clean up.

The main thing I think this class is missing (besides other comments made) is a `geo_point origin() const` function to retrieve the origin as a geo point.

Reply:
Ok, that sounds good, I think we have local_cartesian in shape for this from the discussions below.

Feel free to ping me when you want to tackle the local_geo_cs